pineappl/src/evolution.rs

```diff
@@ -368,6 +425,127 @@ pub(crate) fn ndarray_from_subgrid_orders(
     Ok((fac1, x1_a, x1_b, array))
 }
 
+type X1aX1bOp3Tuple = (Vec<f64>, Vec<f64>, Array3<f64>);
+
+fn ndarray_from_subgrid_orders_slice(
+    info: &OperatorSliceInfo,
+    subgrids: &ArrayView1<SubgridEnum>,
+    orders: &[Order],
+    order_mask: &[bool],
+) -> Result<X1aX1bOp3Tuple, GridError> {
+    // TODO: skip empty subgrids
+
+    let fac1 = info.xif * info.xif * info.fac1;
+    let mut x1_a: Vec<_> = subgrids
+        .iter()
+        .enumerate()
+        .filter(|(index, _)| order_mask.get(*index).copied().unwrap_or(true))
+        .flat_map(|(_, subgrid)| subgrid.x1_grid().into_owned())
+        .collect();
+    let mut x1_b: Vec<_> = subgrids
+        .iter()
+        .enumerate()
+        .filter(|(index, _)| order_mask.get(*index).copied().unwrap_or(true))
+        .flat_map(|(_, subgrid)| subgrid.x2_grid().into_owned())
+        .collect();
+
+    x1_a.sort_by(f64::total_cmp);
+    x1_a.dedup_by(|a, b| approx_eq!(f64, *a, *b, ulps = EVOLUTION_TOL_ULPS));
+    x1_b.sort_by(f64::total_cmp);
+    x1_b.dedup_by(|a, b| approx_eq!(f64, *a, *b, ulps = EVOLUTION_TOL_ULPS));
+
+    let mut array = Array3::<f64>::zeros((1, x1_a.len(), x1_b.len()));
+
+    // add subgrids for different orders, but the same bin and lumi, using the right
+    // couplings
+    for (subgrid, order) in subgrids
+        .iter()
+        .zip(orders.iter())
+        .zip(order_mask.iter().chain(iter::repeat(&true)))
+        .filter_map(|((subgrid, order), &enabled)| {
+            (enabled && !subgrid.is_empty()).then_some((subgrid, order))
+        })
+    {
+        let mut logs = 1.0;
+
+        if order.logxir > 0 {
+            if approx_eq!(f64, info.xir, 1.0, ulps = 4) {
+                continue;
+            }
+
+            logs *= (info.xir * info.xir).ln();
+        }
+
+        if order.logxif > 0 {
+            if approx_eq!(f64, info.xif, 1.0, ulps = 4) {
+                continue;
+            }
+
+            logs *= (info.xif * info.xif).ln();
+        }
+
+        // TODO: use `try_collect` once stabilized
+        let xa_indices: Vec<_> = subgrid
+            .x1_grid()
+            .iter()
+            .map(|&xa| {
+                x1_a.iter()
+                    .position(|&x1a| approx_eq!(f64, x1a, xa, ulps = EVOLUTION_TOL_ULPS))
+                    .ok_or_else(|| {
+                        GridError::EvolutionFailure(format!("no operator for x1 = {xa} found"))
+                    })
+            })
+            .collect::<Result<_, _>>()?;
+        let xb_indices: Vec<_> = subgrid
+            .x2_grid()
+            .iter()
+            .map(|&xb| {
+                x1_b.iter()
+                    .position(|&x1b| approx_eq!(f64, x1b, xb, ulps = EVOLUTION_TOL_ULPS))
+                    .ok_or_else(|| {
+                        GridError::EvolutionFailure(format!("no operator for x1 = {xb} found"))
```

Comment:
```suggestion
                        GridError::EvolutionFailure(format!("no operator for x2 = {xb} found"))
```
if you keep the `2` you can distinguish "flying left" and "flying right" aka 1 and 2 aka `a` and `b` (or else maybe simply drop it) (referring to the above error message)

Reply:
I'm happy if you improve even further, but for now I think you can afford a 1 character change :upside_down_face: (since the line is in the diff any ways)